File: frontend/src/lit/simple-greeting.ts
import {customElement, html, LitElement, property, query} from "lit-element";
import '@polymer/paper-input/paper-input.js';
import {PaperInputElement} from "@polymer/paper-input/paper-input";

@customElement('simple-greeting')
export class SimpleGreeting extends LitElement {

    @property()
    greeting = "";

    @query('#input')
    _input!: PaperInputElement;

    render() {
        return html`
            <div>
                <paper-input id="input"></paper-input>
                <button @click="${this._sayHello}">Say hello</button>
                <div>${this.greeting}</div>
            </div>`;
    }

    _sayHello() {
        const event = new CustomEvent('say-hello', {
            detail: {
                name: this._input.value
            }
        });
        this.dispatchEvent(event);
    }
}

File: src/main/java/org/acme/servlet/lit/SimpleGreeting.java
package org.acme.servlet.lit;

import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.dependency.JsModule;
import com.vaadin.flow.component.dependency.NpmPackage;
import com.vaadin.flow.dom.DomEventListener;

/**
 * Demoes a very simple LitElement. The element fires a <code>say-hello</code>
 * event once the button is clicked. The event is captured by the server-side code,
 * which then sets the <code>greeting<code> property of the client-side component.
 * Lit will then make sure to display the new value of the property.
 * @author Martin Vysny <mavi@vaadin.com>
 */
@Tag("simple-greeting")
@NpmPackage(value = "@polymer/paper-input", version = "3.0.2")
@JsModule("./src/lit/simple-greeting.ts")
public class SimpleGreeting extends Component {
    public SimpleGreeting() {
        getElement().addEventListener("say-hello", (DomEventListener) event -> {
            getElement().setProperty("greeting", "Hello, " + event.getEventData().getString("event.detail.name"));
        }).addEventData("event.detail.name");
    }
}
